homeassistant/components/zwave/__init__.py

```diff
@@ -592,25 +594,38 @@ def start_zwave(_service_or_event):
         network.start()
         hass.bus.fire(const.EVENT_NETWORK_START)
 
-        # Need to be in STATE_AWAKED before talking to nodes.
-        # Wait up to NETWORK_READY_WAIT_SECS seconds for the zwave network
-        # to be ready.
-        for i in range(const.NETWORK_READY_WAIT_SECS):
+        start_time = dt_util.utcnow()
+
+        def _check_awaked(now):
+            """Wait for Z-wave awaked state (or timeout) and finalize start."""
             _LOGGER.debug(
                 "network state: %d %s", network.state,
                 network.state_str)
+
+            waited = int((dt_util.utcnow()-start_time).total_seconds())
+
             if network.state >= network.STATE_AWAKED:
```

Comment:
Does this check do I/O or would it be possible to run this method inside a coroutine? In that case you would be able to just use `yield from asyncio.sleep(1)` instead of scheduling yourself.

Reply:
I think [you are on to something](https://github.com/OpenZWave/python-openzwave/blob/566f1c8ae13cac575bee78dabcd9d46d222a312a/src-api/openzwave/network.py#L493).

So we should revert my rework and just change `time.sleep` to `asyncio.sleep`.

However, I did try that and I found out that I am unsure how to make such a mixed sync/async setup (especially in tests) so it would be great if someone else could contribute that part.

Reply:
You'd want to keep the work you have splitting this into a separate function, and annotate `_check_awaked` with `@asyncio.coroutine`. Then, instead of calling `_check_awaked` directly from `setup`, you'd need to schedule it with `hass.add_job(_check_awaked)`. Finally, unless `_finalize_start` could also be converted, you'd need to schedule that as well to run it in a thread with `hass.async_add_job(_finalize_start)`. I'm not sure about the polling commands, but it seems safer to assume those might contain IO unless we want to dive into it.

Reply:
I have pushed a commit changing it like that. However, I had some trouble with the `asyncio.sleep` mock and still think that it is clumsy.